host/crypto/windows/rsa.c

```diff
@@ -0,0 +1,556 @@
+    buffer_local = NULL;
+    result = OE_OK;
+
+done:
+    if (buffer_local)
+    {
+        oe_secure_zero_fill(buffer_local, buffer_local_size);
+        free(buffer_local);
+        buffer_local_size = 0;
+    }
+
+    return result;
+}
+
+oe_result_t oe_rsa_get_public_key_from_private(
+    const oe_rsa_private_key_t* private_key,
+    oe_rsa_public_key_t* public_key)
+{
+    oe_result_t result = OE_UNEXPECTED;
+    oe_private_key_t* impl = (oe_private_key_t*)private_key;
+    uint8_t* keybuf = NULL;
+    ULONG keybuf_size;
+    BCRYPT_RSAKEY_BLOB* key_header;
+
+    if (!oe_private_key_is_valid(impl) || !public_key)
+        OE_RAISE(OE_INVALID_PARAMETER);
+
+    OE_CHECK(_get_public_rsa_blob_info(impl->pkey, &keybuf, &keybuf_size));
+
+    /* Sanity check to ensure we get modulus and public exponent. */
+    key_header = (BCRYPT_RSAKEY_BLOB*)keybuf;
+    if (key_header->cbPublicExp == 0 || key_header->cbModulus == 0)
+        OE_RAISE(OE_FAILURE);
+
+    /*
+     * Export the key blob to a public key. Note that the private key blob has
+     * the modulus and the exponent already, so we can just use it to import
+     * the public key.
+     */
+    {
+        NTSTATUS status;
+        BCRYPT_KEY_HANDLE public_key_handle;
+
+        status = BCryptImportKeyPair(
+            BCRYPT_RSA_ALG_HANDLE,
+            NULL,
+            BCRYPT_RSAPUBLIC_BLOB,
+            &public_key_handle,
+            keybuf,
+            keybuf_size,
+            0);
+
+        if (status != STATUS_SUCCESS)
+            OE_RAISE(OE_FAILURE);
+
+        ((oe_public_key_t*)public_key)->magic = _PUBLIC_KEY_MAGIC;
+        ((oe_public_key_t*)public_key)->pkey = public_key_handle;
+    }
+
+    result = OE_OK;
+
+done:
+    if (keybuf)
+    {
+        oe_secure_zero_fill(keybuf, keybuf_size);
+        free(keybuf);
+        keybuf_size = 0;
+    }
+
+    return result;
+}
+
+oe_result_t oe_rsa_public_key_get_modulus(
+    const oe_rsa_public_key_t* public_key,
+    uint8_t* buffer,
+    size_t* buffer_size)
+{
+    oe_result_t result = OE_UNEXPECTED;
+    const oe_public_key_t* impl = (const oe_public_key_t*)public_key;
+    uint8_t* keybuf = NULL;
+    ULONG keybuf_size;
+    BCRYPT_RSAKEY_BLOB* keyblob;
+
+    /* Check for null parameters and invalid sizes. */
+    if (!oe_public_key_is_valid(impl) || !buffer_size ||
+        *buffer_size > MAXDWORD)
+    {
+        OE_RAISE(OE_INVALID_PARAMETER);
+    }
+
+    /* If buffer is null, then buffer_size must be zero */
+    if (!buffer && *buffer_size != 0)
+        OE_RAISE(OE_INVALID_PARAMETER);
+
+    OE_CHECK(_get_public_rsa_blob_info(impl->pkey, &keybuf, &keybuf_size));
+
+    keyblob = (BCRYPT_RSAKEY_BLOB*)keybuf;
+    if (keyblob->cbModulus > *buffer_size)
+    {
+        *buffer_size = keyblob->cbModulus;
+        OE_RAISE(OE_BUFFER_TOO_SMALL);
+    }
+
+    OE_CHECK(
+        oe_memcpy_s(
+            buffer,
+            *buffer_size,
+            keybuf + sizeof(*keyblob) + keyblob->cbPublicExp,
```

Comment:
Probably worth a comment to describe the `BCRYPT_RSAKEY_BLOB` as being the header concatenated with the public exp and modulus in that order, here and in `oe_rsa_public_key_get_exponent`.

Reply:
sounds good.